File: code_snippets/ts/src/use_cases/workflows/subscription_workflow.ts
import * as restate from "@restatedev/restate-sdk/lambda";

// <start_here>
const subscriptionWorkflow = restate.workflow({
  name: "SubscriptionWorkflow",
  handlers: {
    run: async (ctx: restate.WorkflowContext, req: SubscriptionRequest) => {
      // <mark_2> orange
      const compensations = [];
      // </mark_2>

      try {
        const paymentId = ctx.rand.uuidv4();
        // <mark_2> orange
        compensations.push(() => removeRecurringPayment(paymentId))
        // </mark_2>
        // <mark_2> green
        await ctx.run(() => createRecurringPayment(req.creditCard, paymentId));
        // </mark_2>
        // <mark_1>
        ctx.set("status", "payment_success");
        // </mark_1>

        for (const subscription of req.subscriptions) {
          // <mark_2> orange
          compensations.push(() => removeSubscription(req.userId, subscription))
          // </mark_2>
          // <mark_2> green
          await ctx.run(() => createSubscription(req.userId, subscription));
          // </mark_2>
        }
        // <mark_1>
        ctx.set("status", "subscribed");
        // </mark_1>
      } catch (e) {
        // <mark_2> orange
        if (e instanceof restate.TerminalError) {
          for (const compensation of compensations.reverse()) {
            await ctx.run(() => compensation());
          }
          // </mark_2>
          // <mark_1>
          ctx.set("status", "rolled_back");
          // </mark_1>
        }
        throw e;
      }
    },
  },
})

// <mark_3>
export const awsLambdaHandler = restate
    .endpoint()
    .bind(subscriptionWorkflow)
    .handler();
// </mark_3>
// <end_here>


type SubscriptionRequest = {creditCard: string, userId: any, subscriptions: any[]};

async function removeRecurringPayment(paymentId: string){}
async function createRecurringPayment(creditCard: string, paymentId: string){}

async function removeSubscription(userId: any, subscription: any, payRef: void) {}
async function createSubscription(userId: any, subscription: any, payRef: void) {
  return undefined;
}

File: code_snippets/python/src/use_cases/workflows/subscription_workflow.py
import uuid

import restate
from pydantic import BaseModel
from restate import Workflow, WorkflowContext, WorkflowSharedContext
from restate.exceptions import TerminalError


class SubscriptionRequest(BaseModel):
    user_id: str
    credit_card: str
    subscriptions: list[str]


# <start_here>
subscription_workflow = Workflow("SubscriptionWorkflow")


@subscription_workflow.main()
async def run(ctx: WorkflowContext, req: SubscriptionRequest):
    # <mark_2> orange
    compensations = []
    # </mark_2>

    try:
        payment_id = await ctx.run("payment id", lambda: str(uuid.uuid4()))

        # <mark_2> orange
        compensations.append(lambda: remove_recurring_payment(payment_id))
        # </mark_2>
        # <mark_2> green
        await ctx.run("recurring payment",
                            lambda: create_recurring_payment(req.credit_card, payment_id))
        # </mark_2>
        # <mark_1>
        ctx.set("status", "payment_success");
        # </mark_1>

        for subscription in req.subscriptions:
            # <mark_2> orange
            compensations.append(lambda: remove_subscription(req.user_id, subscription))
            # </mark_2>
            # <mark_2> green
            await ctx.run("subscription",
                      lambda: create_subscription(req.user_id, subscription))
            # </mark_2>
        # <mark_1>
        ctx.set("status", "subscribed");
        # </mark_1>

        # <mark_2> orange
    except TerminalError as e:
        for compensation in reversed(compensations):
            await ctx.run("run compensation", compensation)
            # </mark_2>
        # <mark_1>
        ctx.set("status", "rolled_back");
        # </mark_1>
        raise e


# <mark_3>
aws_lambda_handler = restate.app([subscription_workflow])
# </mark_3>
# <end_here>

def create_recurring_payment(credit_card, payment_id):
    return None


def remove_recurring_payment(payment_id):
    pass


def create_subscription(user_id, subscription, pay_ref):
    return None

def remove_subscription(user_id, subscription):
    return None